### ISSUE TITLE
Research: SIGReg-assisted population routing calibration—test the precise need for Winning

### ISSUE THREAD

**Issue by microprediction** [microprediction/winning#65]
The useful connection to SIGReg/LeJEPA appears narrower than “Gaussian embeddings + a Winning output layer.” Proposed research: **use Winning offline to calibrate population routing shares for an already-justified noisy top-1 gate whose input factors are approximately Gaussian.** Establish whether accurate, repeated share inversion earns its cost over cheap sampling and load feedback before building an integration.

This is a research hypothesis, not a claim of improved representation learning or a request to change the production API.

**Why ordinary LeJEPA does not need Winning.** For a fixed input with deterministic features and scores, choosing the largest score is just an argmax. SIGReg’s distribution-matching loss also needs no race computation. Moreover, in the authors’ [minimal implementation](https://github.com/galilai-group/lejepa/blob/c293d291ca87cd4fddee9d3fffe4e914c7272052/MINIMAL.md), SIGReg is applied to the projector output, while the probe consumes backbone embeddings. A Gaussian projector output does not establish Gaussianity of the probe features. The [paper](https://arxiv.org/abs/2511.08544) concerns a marginal distribution across inputs/views; it does not supply a Gaussian posterior for any particular input.

The proposed application deliberately uses that *population* distribution, avoiding this conditional-uncertainty inference.

**The concrete operation.** Suppose a stochastic gate, with a small routing bottleneck, is already appropriate:

$$
F=f_\theta(X)\approx N(0,I_r),\qquad
S_i=a_i^\top F+\sigma_i\epsilon_i,\qquad
R_b=\arg\max_i(S_i+b_i),
$$

where the $\epsilon_i$ are independent standard Gaussians, independent of $F$, and $A,\sigma$ are fixed during calibration. The gate chooses one expert; the experts can still receive the full backbone representation. Apply SIGReg to the actual routing factors $F$, and retain a task objective so their informativeness is measured.

The desired quantity is

$$
p_i(b)=\Pr_{X,\epsilon}(R_b=i),
\qquad p(b)=q,
$$

where $q$ is a specified positive vector of long-run allocation fractions, summing to one. It need not be uniform.

Under the ideal Gaussian factor law this is exactly the factor-probit *model* with covariance $AA^\top+\operatorname{diag}(\sigma^2)$. Winning supplies the numerical forward map and its inverse. Calibrate biases between routing deployments or configuration changes, then route each actual input with the ordinary score calculation and argmax. **There is no Winning call per input.**

This is also a familiar convex dual: $H(b)=E[\max_i(S_i+b_i)]-q^\top b$ has gradient $p(b)-q$. A common bias is unidentified, so fix $\sum_i b_i=0$. The contribution would be evaluating/inverting that gradient efficiently for the structured Gaussian law; the balancing principle itself is not new. Stochastic semi-discrete transport already offers sample-based dual optimization: [Genevay et al., 2016](https://arxiv.org/abs/1605.08527).

The independent noise must be part of the declared routing policy, or represent justified independent score residuals. **Do not silently add jitter solely to fit the solver.** General deterministic affine scores on Gaussian embeddings instead have covariance $AA^\top$, potentially high rank and singular; their conditional races can become deterministic. The proposed smooth factor-plus-diagonal experiment does not establish that harder case. Input-dependent noise scales also require further integration rather than being replaced by one constant $D$.

**Why correlation can change the answer.** With zero biases, take

$$
A=(\sqrt{0.96},\sqrt{0.96},0)^\top,\qquad
D=(0.04,0.04,1).
$$

All three scores have mean zero and variance one, but the first two have correlation $0.96$. Their population winning shares are

$$
(0.26594214,\ 0.26594214,\ 0.46811572),
$$

not one third each. The third share has the independent check $1/4+\arcsin(0.98)/(2\pi)$. Winning recovers mean-zero biases approximately

$$
b=(0.16480057,\ 0.16480057,\ -0.32960114)
$$

for equal target shares. Independent one-dimensional integration of each bivariate difference event verified the resulting shares to a maximum absolute discrepancy of $2.1\times10^{-10}$ from $1/3$.

Reproduction using the current API:

~~~python
import numpy as np
from winning.probit import shares
from winning.factor.races import abilities_from_race

A = np.array([[np.sqrt(.96)], [np.sqrt(.96)], [0.]])
D = np.array([.04, .04, 1.])  # variances, not standard deviations
q = np.full(3, 1 / 3)

print(shares(np.zeros(3), V=A, D=D, points=1001))
mu, info = abilities_from_race(
    q, V=A, D=D, points=1001,
    tol=1e-9, n_iter=120, return_info=True,
)
assert info["converged"], info
b = -mu  # internal API uses min-wins locations
print(b)
print(shares(b, V=A, D=D, points=1001))
~~~

This three-alternative example demonstrates the quantity that must be calculated; it is not a performance argument for using Winning at that size.

**The strongest simple alternative must be included.** Crude Monte Carlo would require approximately $(1-p)/(p\eta^2)$ independent draws for relative standard error $\eta$ on a share $p$. For $p=10^{-5}$ and $\eta=0.1$, that is about $10^7$ draws. But that observation alone is an inadequate justification for Winning.

[Shazeer et al., 2017, Appendix A](https://arxiv.org/html/1701.06538v1#A1) integrates out one expert’s Gaussian noise while conditioning on the others. Adapted to this model, draw $F$ and the other shocks and estimate each share with

$$
\Phi\!\left(
\frac{b_i+a_i^\top F-\max_{j\ne i}(b_j+a_j^\top F+\sigma_j\epsilon_j)}
{\sigma_i}
\right).
$$

The largest and second-largest sampled scores supply all leave-one-out thresholds in $O(n)$ per draw. This estimator can return tiny probabilities without observing a rare win. It is unbiased before optional finite-sample normalization. It is an essential $O(Mnr)$ competitor, not an $O(Mn^2)$ per-alternative integration straw man.

Additional baselines:

- Empirical bias feedback, including sign updates inspired by [Wang et al., 2024](https://arxiv.org/abs/2408.15664). Their published gate adds biases to transformed affinity scores; an affine-score implementation here is an adaptation, not an exact reproduction.
- Direct Gaussian Monte Carlo and scrambled Sobol, with common random draws for repeated bias evaluations. Also combine scrambled Sobol with the conditional-CDF estimator; raw QMC alone is not the strongest sampling baseline.
- Sample-based dual optimization on replayed real factors; this can succeed when the Gaussian model is wrong.
- [SwAV](https://arxiv.org/abs/2006.09882) and [BASE](https://arxiv.org/abs/2103.16716) for context: batch assignments and exact batch capacities are different objectives. The proposed population calibration does not replace their batch guarantees.

**Initial measurements, not a completed benchmark.** I ran the Python implementation at commit [5c4bdff](https://github.com/microprediction/winning/tree/5c4bdff7ed944a0c9552181e44f5fb57829126d6), without Rust, using one OpenBLAS thread:

- $n=128$, rank one, $A_i$ from NumPy default_rng(1709).uniform(-.75,.75,(128,1)), $D_i=.75$, $q_0=10^{-5}$, and the other targets equal to $(1-10^{-5})/127$.
- Default 15 factor nodes / 257 lattice points: forward evaluation about 21 ms; inversion about 0.50 s and 21 iterations. The internal relative target residual was $4.5\times10^{-9}$, **but this is not the integration error**. Repricing with 63 factor nodes / 1025 points moved the worst share by $4.1\times10^{-7}$ absolute, or $5.2\times10^{-5}$ relative. Refining only the lattice had concealed this factor-quadrature error.
- The normalized conditional-CDF estimator, over five seeds (500–504), took approximately 4, 17, and 65 ms at 1,024 / 4,096 / 16,384 draws. Median total-variation errors against the refined numerical reference were approximately 0.0126 / 0.00657 / 0.00326. At 4,096 draws the rare-share relative errors ranged from 0.0091 to 0.0765. This already defeats a claim that all sampling requires observing rare wins.

These numbers support investigating a **high-accuracy, whole-vector, repeated-inversion niche**. They do not establish an advantage for rough load balancing, GPU training, or real SIGReg features. A single configuration and a refined reference are not an accuracy certificate or a general speed comparison.

**Where SIGReg earns its place, if anywhere.** It must reduce *population model error*, rather than merely improve the numerical solver’s internal residual. In the three-score example above, replacing the Gaussian factor by a symmetric $\{-1,+1\}$ factor preserves all score means and covariances. Keeping the Gaussian-calibrated biases nevertheless produced empirical shares approximately $(0.3249,0.3247,0.3505)$ over one million draws. Whitening alone therefore does not justify the Gaussian occupancy calculation.

Conversely, small SIGReg loss on finitely many projections does not certify accurate masses for all winning regions or relative accuracy in rare tails. At fixed routing policy, total variation contracts from the factor law to the choice law, but a useful bound on that factor-law discrepancy is not provided by merely reporting the training loss. Validate the actual route shares.

**A bounded experiment and stopping rule.**

1. First benchmark the numerical operation with known Gaussian factors: $n\in\{32,128,512\}$ and $r\in\{1,2,4\}$, heterogeneous positive target shares, clustered loadings, and several meaningful residual-noise levels. Include a sharp/small-noise stress case. Measure all-share error, tail relative error, bias-inversion convergence, total calibration time and a counterfactual expert removal. Compare at matched accuracy, not just matched draw counts. Vary both factor-node and lattice budgets.
2. Benchmark the conditional-CDF estimator, raw MC/RQMC, stochastic dual updates, and load feedback at the same targets. For inversion include the cost of every probability estimate and tuning pass. Use multiple seeds and report uncertainty. Exact quota satisfaction for an observed batch is out of scope.
3. Only if stage 1 identifies a useful regime, test a small learned routing bottleneck with the same task loss and three regularizers: none, covariance whitening, SIGReg. Keep routing dimension, noise policy, optimizer budget and task evaluation comparable. Compare actual held-out allocation fractions against both Gaussian predictions and empirical calibration. Include distribution shift and retained backbone features as a negative control. Report task quality alongside share accuracy. A small SIGReg value on its own is not a success metric.
4. Separate solver error, Gaussian-law approximation error and deployment shift. Do not impose uniform class frequencies on a naturally imbalanced classification task, or treat a calibrated allocation fraction as probability of semantic correctness.

A proposed *numerical* continuation threshold is maximum relative share error below 0.01 on declared targets down to $10^{-5}$, with materially lower total inversion cost than the best baseline. This threshold is a research design choice, not an application requirement. Before a real integration, identify an actual consumer that needs such precision and can validate the population model. **Stop if cheaper feedback/CDF estimates meet that consumer’s tolerance, if the required factor rank/noise regime erases the speed advantage, or if Gaussianization harms useful routing or leaves model error larger than the claimed gain.**

Existing [forward/inverse APIs](https://github.com/microprediction/winning/blob/5c4bdff7ed944a0c9552181e44f5fb57829126d6/winning/factor/races.py) suffice for this experiment. No new autograd layer is required. Bias derivatives exist in [polish.py](https://github.com/microprediction/winning/blob/5c4bdff7ed944a0c9552181e44f5fb57829126d6/winning/factor/polish.py); they differentiate the normalized computation conditional on the selected grid and omit grid motion. Do not present them as a complete end-to-end training interface for learned loadings and noise scales.

Finally, the existing [probit-output-layer note](https://github.com/microprediction/winning/blob/5c4bdff7ed944a0c9552181e44f5fb57829126d6/research/applications/probit_output_layer.md) concerns a separate conditional-noise application. [HET/HET-XL](https://arxiv.org/abs/2301.12860) explicitly models conditional Gaussian noise and approximates expectations with sampling; this does not require SIGReg, and a finite-temperature softmax expectation is not the same target as a hard Gaussian argmax probability. Keep that investigation separate.

Requested deliverable: one reproducible research script and a short result table that either establishes this specific calibration advantage or records that the connection is mathematically valid but operationally unnecessary.

**Comment by microprediction** [comment]
Follow-up from the discussion: a more concrete application is a **batched correlated combinatorial bandit whose action is a binary portfolio of routes**. For example, a soccer news feed sends thousands of messages and learns which channels to keep using from average receipt feedback. It estimates route effectiveness and shared failure structure, then selects a binary portfolio for subsequent messages. SIGReg is an optional way to support the score-distribution model; the bandit application stands independently.

**Action and objective.** Let \(a\in\{0,1\}^n\), with \(c^\top a\le B\) and optionally \(\sum_i a_i\le k\). Under latent route scores \(X_i\) and an outside score \(X_0\), the portfolio reward is

$$
R(a)=\mathbf 1\{\max_{i:a_i=1}X_i>X_0\},\qquad
q_\theta(a)=E_\theta[R(a)].
$$

Unselected routes do not run. This is the group-entry objective already described in [research/selection/NOTES.md](https://github.com/microprediction/winning/blob/5c4bdff7ed944a0c9552181e44f5fb57829126d6/research/selection/NOTES.md). The additional work is learning from portfolio-level feedback and exploring informative portfolios.

The nonlinear receipt objective is essential. If the reward were simply the sum of selected channels' rewards and the objective its expectation, dependence would not change the optimal portfolio at known marginal means. Likewise, selecting a set to contain the winner of an unchanged all-channel race has the modular objective \(\sum_{i\in S}p_i\). Neither is the proposed complementary-route selection problem.

**What steady-state averages retain.** For independent, stationary trials using the same portfolio \(S\), receipt count \(K_S\) and exposure count \(N_S\) give

$$
K_S\sim\mathrm{Binomial}(N_S,q_\theta(S)).
$$

These counts retain the Bernoulli likelihood; individual trial order is unnecessary under those assumptions. Preserve portfolio identity and exposure count. An average pooled across different portfolios, or separate channel marginal success averages, does not generally identify joint failure probabilities. Serial dependence requires a different uncertainty calculation.

If the feedback instead consists of winner counts by channel within known choice sets, use a multinomial race likelihood. Existing share inversion can estimate abilities conditional on a specified dependence model, but one vector of winning shares does not identify that model's covariance.

**Where Winning contributes.** For a factor model \(X_i=\mu_i+v_i^\top F+\sigma_i\epsilon_i\), with independent residuals conditional on \(F\), and fixed outside threshold, let \(p_i(F)\) be conditional receipt probability. Then

$$
q_\theta(S)=1-E_F\prod_{i\in S}(1-p_i(F)),
$$

and every candidate addition has value

$$
\Delta(j\mid S)=E_F\left[p_j(F)\prod_{i\in S}(1-p_i(F))\right].
$$

A selected portfolio's conditional failure product is reusable across all candidates. Random outside scores add an integration over the outside score, conditional on shared factors. This is the shared-field/group-selection computation, rather than a requirement to calculate every individual winning probability first. A fixed-threshold objective is simpler than full winner-share integration and should be benchmarked accordingly.

Dependence matters because an individually strong route may almost always succeed and fail alongside routes already selected. The useful quantity is its incremental chance of receipt when those routes fail. Under fixed potential route outcomes, this coverage objective is monotone submodular; the existing note develops greedy marginal-gain evaluation.

**A simpler learning loop: epsilon-greedy feed routing.**

1. For a small fraction \(\epsilon\) of messages, choose a budget-feasible channel portfolio at random, independently of the message's expected appeal. For the remaining messages, use the current best portfolio under the fitted model.
2. Record the selected set, its selection probability, exposure count, receipt feedback, and time window. Fit the model periodically using the random exploration traffic; the exploration flag allows a clean initial estimation sample.
3. Recompute the best portfolio and continue. The race distribution is over typical soccer messages and receiver states; this first experiment does not require a separate uncertainty model for every message.
4. If deployment sends to three channels, exploration should include varied triples. Singleton tests reveal marginal receipt rates but cannot identify joint failure/redundancy from separate averages alone.

This deliberately uses random exploration rather than requiring posterior sampling or an elaborate exploration optimizer. The factor model shares information between tested and untested portfolios; randomization supplies evidence, while generalization to untried sets remains a model assumption to validate. Averages should remain indexed by selected portfolio (and relevant context/time window), with exposure counts retained.

Start with specified or strongly constrained dependence and unknown abilities; then test whether dependence can be learned from the available varied portfolios. The aggregate-feedback estimator and exploration policy are additional components; ordinary shares-to-abilities inversion alone is insufficient.

**Connection to multi-prototype routing.** If Gaussian query embeddings are scored against fixed paraphrase vectors, the primitive inner products are jointly Gaussian. A route scored by its maximum paraphrase score is a group maximum: with ties resolved, its winning probability is the sum of its constituent prototypes' global winning probabilities. This can supply a structured dependence hypothesis. Low rank within each route does not guarantee low global rank; block/nested structure must also be checked. Squared subspace-projection scores are quadratic forms rather than Gaussian scores. SIGReg's population Gaussianity does not establish conditional receipt uncertainty.

**Bounded experiment.** First use known low-rank/block dependence and unknown abilities, collect only batched aggregate receipt feedback, and test recovery of held-out portfolio probabilities and cumulative receipts under a fixed budget. Compare with (i) an independence model, (ii) a generic nonlinear full-bandit baseline, and (iii) the same structured learning policy with MC/RQMC probability calculations. Then relax the known-dependence assumption. Report selection quality, uncertainty calibration, computational cost, and robustness to misspecified factors; do not infer a learning advantage solely from faster numerical evaluation.

Relevant prior work: [Nie et al., ICML 2023](https://arxiv.org/abs/2301.13326) studies nonlinear combinatorial rewards with bandit feedback and adapting offline approximation algorithms. [Rejwan and Mansour, 2019](https://arxiv.org/abs/1905.12624) studies full-bandit feedback for additive top-k rewards, an important distinct objective.

The specific hypothesis is that a structured race model lets observations on some portfolios predict the receipt probability of untried portfolios, while Winning makes repeated counterfactual selection economical. A decisive test must establish both the statistical transfer and the computational benefit.
